params/precompile_config.go

```diff
@@ -0,0 +1,281 @@
+func (c *ChainConfig) checkPrecompileCompatible(key precompileKey, precompileUpgrades []PrecompileUpgrade, headTimestamp *big.Int) *ConfigCompatError {
+	// all active upgrades must match
+	activeUpgrades := c.getActivatingPrecompileConfigs(nil, headTimestamp, key, c.PrecompileUpgrades)
+	newUpgrades := c.getActivatingPrecompileConfigs(nil, headTimestamp, key, precompileUpgrades)
+
+	// first, check existing upgrades are there
+	for i, upgrade := range activeUpgrades {
+		if len(newUpgrades) <= i {
+			// missing upgrade
+			return newCompatError(
+				fmt.Sprintf("missing PrecompileUpgrade[%d]", i),
+				upgrade.Timestamp(),
+				nil,
+			)
+		}
+		// All upgrades that have forked must be identical.
+		if !upgrade.Equal(newUpgrades[i]) {
+			return newCompatError(
+				fmt.Sprintf("PrecompileUpgrade[%d]", i),
+				upgrade.Timestamp(),
+				newUpgrades[i].Timestamp(),
+			)
+		}
+	}
+	// then, make sure newUpgrades does not have additional upgrades
+	// that are already activated. (cannot perform retroactive upgrade)
+	if len(newUpgrades) > len(activeUpgrades) {
+		return newCompatError(
+			fmt.Sprintf("cannot retroactively enable PrecompileUpgrade[%d]", len(activeUpgrades)),
+			nil,
+			newUpgrades[len(activeUpgrades)].Timestamp(), // this indexes to the first element in newUpgrades after the end of activeUpgrades
+		)
+	}
+
+	return nil
+}
+
+// EnabledStatefulPrecompiles returns a slice of stateful precompile configs that
+// have been activated through an upgrade.
+func (c *ChainConfig) EnabledStatefulPrecompiles(blockTimestamp *big.Int) []precompile.StatefulPrecompileConfig {
+	statefulPrecompileConfigs := make([]precompile.StatefulPrecompileConfig, 0)
+	for _, key := range precompileKeys {
+		if config := c.getActivePrecompileConfig(blockTimestamp, key, c.PrecompileUpgrades); config != nil {
+			statefulPrecompileConfigs = append(statefulPrecompileConfigs, config)
+		}
+	}
+
+	return statefulPrecompileConfigs
+}
+
+// CheckConfigurePrecompiles checks if any of the precompiles specified by the chain config are enabled or disabled by the block
+// transition from [parentTimestamp] to the timestamp set in [blockContext]. If this is the case, it calls [Configure]
+// or [Deconfigure] to apply the necessary state transitions for the upgrade.
+// This function is called:
+// - within genesis setup to configure the starting state for precompiles enabled at genesis,
+// - during block processing to update the state before processing the given block.
+func (c *ChainConfig) CheckConfigurePrecompiles(parentTimestamp *big.Int, blockContext precompile.BlockContext, statedb precompile.StateDB) {
+	blockTimestamp := blockContext.Timestamp()
+	for _, key := range precompileKeys { // Note: configure precompiles in a deterministic order.
+		for _, config := range c.getActivatingPrecompileConfigs(parentTimestamp, blockTimestamp, key, c.PrecompileUpgrades) {
```

Comment:
Will this suicide multiple times if multiple blocks are produced at the same timestamp? Or do we enforce a precompile upgrade is only processed once at a timestamp (maybe only the boundary of less than time to >= time)?

Reply:
AFAICT this logic enforces we only run once on the boundary and we should be good.